test/hotspot/jtreg/compiler/lib/verify/Verify.java

```diff
@@ -0,0 +1,302 @@
+ */
+public final class Verify {
+    /**
+     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     */
+    public static void checkEQ(Object a, Object b) {
+        checkEQ(a, b, "");
+    }
+
+    /**
+     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     */
+    private static void checkEQ(Object a, Object b, String context) {
+        // Both null
+        if (a == null && b == null) {
+            return;
+        }
+
+        // Null mismatch
+        if (a == null || b == null) {
+            System.err.println("ERROR: Verify.checkEQ failed: null mismatch");
+            print(a, "a " + context);
+            print(b, "b " + context);
+            throw new VerifyException("Object array null mismatch.");
+        }
+
+        // Class mismatch
+        Class ca = a.getClass();
+        Class cb = b.getClass();
+        if (ca != cb) {
+            System.err.println("ERROR: Verify.checkEQ failed: class mismatch.");
+            System.err.println("       " + ca.getName() + " vs " + cb.getName());
+            print(a, "a " + context);
+            print(b, "b " + context);
+            throw new VerifyException("Object array class mismatch.");
+        }
+
+        switch (a) {
+            case Object[]  x -> checkEQimpl(x, (Object[])b,                context);
+            case Byte      x -> checkEQimpl(x, ((Byte)b).byteValue(),      context);
+            case Character x -> checkEQimpl(x, ((Character)b).charValue(), context);
+            case Short     x -> checkEQimpl(x, ((Short)b).shortValue(),    context);
+            case Integer   x -> checkEQimpl(x, ((Integer)b).intValue(),    context);
+            case Long      x -> checkEQimpl(x, ((Long)b).longValue(),      context);
+            case Float     x -> checkEQimpl(x, ((Float)b).floatValue(),    context);
+            case Double    x -> checkEQimpl(x, ((Double)b).doubleValue(),  context);
+            case byte[]    x -> checkEQimpl(x, (byte[])b,                  context);
+            case char[]    x -> checkEQimpl(x, (char[])b,                  context);
+            case short[]   x -> checkEQimpl(x, (short[])b,                 context);
+            case int[]     x -> checkEQimpl(x, (int[])b,                   context);
+            case long[]    x -> checkEQimpl(x, (long[])b,                  context);
+            case float[]   x -> checkEQimpl(x, (float[])b,                 context);
+            case double[]  x -> checkEQimpl(x, (double[])b,                context);
+            case MemorySegment x -> checkEQimpl(x, (MemorySegment) b,      context);
+            default -> {
+                System.err.println("ERROR: Verify.checkEQ failed: type not supported: " + ca.getName());
+                print(a, "a " + context);
+                print(b, "b " + context);
+                throw new VerifyException("Object array type not supported: " + ca.getName());
```

Comment:
This pattern repeats 3 times here and also below. Maybe you want to consider to move it to a `reportError(String message, String details, a, b, context)` method?

Reply:
It does repeat somewhat, but not really identically. With different types for a, sometimes with context, sometimes without. Feel free to add code suggestions, because I'm not sure what you are exactly imagining.